Feature: add `GovernorVetoGuardian`

funkyenough opened the pull request:
Resolves #7 

- Add `GuardianVetoGuardian.sol`
- Add unit tests in `GuardianVetoGuardian.t.sol`
- Add natspec

funkyenough committed:
feat: add GovernorVetoGuardian

funkyenough committed:
update testis and scripts

funkyenough committed:
chore: rename function guardianVeto to vetoByGuardian

funkyenough committed:
test: add GovernorVetoGuardian.t.sol

thelostone-mc commented on script/DeployTest.s.sol:
```diff
@@ -42,6 +43,7 @@ contract DeployOptimisticGovernance is Script, StdAssertions {
   TimelockController public timelock;
   BasicCouncilVetoGovernor public vetoGovernor;
   BasicCouncilGovernor public councilGovernor;
+  address public vetoGuardian = makeAddr("VetoGuardian");
```
nit: but we could populate this DeployInput File (maybe it's alright for now and we can fix it when we write the deploy scripts)

funkyenough replied:
Definitely will be included in the DeployInput contracts, but for unit tests, maybe we can just mock a regular address.

Alternatively, we can fuzz the veto guardian too.

thelostone-mc commented on src/extensions/GovernorVetoGuardian.sol:
```diff
@@ -0,0 +1,74 @@
+/// @author [ScopeLift](https://scopelift.co)
+/// @notice Extension of {Governor} that allows a trusted address to veto a proposal if the proposal
+/// state is pending or active.
+abstract contract GovernorVetoGuardian is Governor {
```
how is this contract different from [GovernorVetoOverride.sol](https://github.com/withtally/optimistic-governance/pull/8/files#diff-f8b5fbceb9d9e7a0c715bd8be7d4ec59d28151b5586e39308c82e38c57fda21a) Added in https://github.com/withtally/optimistic-governance/pull/8/files
@funkyenough ?

funkyenough replied:
`GovernorVetoOverride` allows the `vetoOverrideRole` (essentially the main DAO) to change the state of a `Defeated` proposal to `Succeeded` under certain conditions.

`GovernorVetoGuardian` is a trusted address that can change a `Pending` or `Active` proposal on the veto governor to `Defeated`.

The naming is quite tricky, and maybe we can come up with a better name

funkyenough committed:
refactor: add _setVetoGuardian internal method, update vetoByGuardian check logic, update natspec

funkyenough committed:
Use veto guardian mock contract instead of smoke test base
remove setVetoGuardian tests
add _setVetoGuardian tests
add state tests
update helper contract

wildmolasses commented on src/BasicCouncilVetoGovernor.sol:
```diff
@@ -109,10 +113,10 @@ contract BasicCouncilVetoGovernor is
     returns (ProposalState)
   {
-    return GovernorVetoOverride.state(proposalId);
+    return super.state(proposalId);
```
let's be more explicit about the inheritance order here. the order is correct, but we should make a comment as to why.

wildmolasses replied:
see #30

wildmolasses commented on src/extensions/GovernorVetoGuardian.sol:
```diff
@@ -0,0 +1,77 @@
+  function _setVetoGuardian(address _vetoGuardian) internal virtual {
+    address _oldVetoGuardian = vetoGuardian;
+    vetoGuardian = _vetoGuardian;
+    emit VetoGuardianModified(_oldVetoGuardian, _vetoGuardian);
```
nit

```suggestion
    emit VetoGuardianSet(_oldVetoGuardian, _vetoGuardian);
```

wildmolasses commented on src/extensions/GovernorVetoGuardian.sol:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  /// @notice Set the veto guardian address. Can only be called by the governance.
+  /// @dev Passing address(0) effectively removes the guardian. Callers must ensure that is
```
nit: can remove the note about the callers

```suggestion
  /// @dev Passing address(0) effectively removes the guardian.
```

wildmolasses commented on test/GovernorVetoGuardian.t.sol:
```diff
@@ -0,0 +1,401 @@
+  ) public returns (uint256 _proposalId) {
+    _proposalId = _passAndQueueProposal(_proposer, _caller, _proposal);
+
+    vm.warp(block.timestamp + 1 days + 1);
```
nit: refactor out the magic numbers

wildmolasses replied:
#31
